FsaMatch.php

```diff
@@ -0,0 +1,91 @@
+<?php
```

Comment:
https://github.com/exussum12/fsamatch/blob/master/src/FsaMatch.php  This comes from here, its fully unit tested

Reply:
That's pretty cool - how long did it take to scrape the FSA data? My initial thought is that we'd probably have to do this fairly regularly.

Reply:
About 40 minutes single threaded. Could probably speed it up. 

Normally yes but given the current situation I'm not sure FSA are going out.  I can set up a GitHub action to keep that file up to date if needed   I'll post the script too to the same repo

Reply:
Nice - maybe worth setting up a Github action to poll it every couple of weeks!

Reply:
Is there any way we could reformat the rating images a little - I think they're very prominent. I'm maybe we could of making them a little smaller and possibly take off the word "Category" in the bubble

Reply:
Not 100% sure on licence. I would imagine so. Not home currently but I'll check the download pack when I'm back. Want it smaller on both the popup and the list page ?

Reply:
According to https://www.food.gov.uk/terms-and-conditions the images can be used with needing to ask.
On the images available here https://ratings.food.gov.uk/open-data-resources/images/images.zip (linked from https://ratings.food.gov.uk/open-data/en-gb) - I would suggest using the `small/72ppi` images for the popups on the map.  

![fhrs_2_en-gb](https://user-images.githubusercontent.com/11561320/89426798-6e794780-d732-11ea-94f8-61e7b2c7c12d.jpg)

And the medium 72ppi ones for the list:

![fhrs_4_en-gb](https://user-images.githubusercontent.com/11561320/89426842-7933dc80-d732-11ea-9d5b-50a35d0d6d18.jpg)

Reply:
The small 72 ppi are are the ones used at the current size, (ie not resized with css). I know they were usable wasn't sure if you could resize etc or had to be used as is.

Ill add the medium ones when i'm back at a computer later on (or feel free to edit the PR in the mean time)

Reply:
And if that still looks a bit too big then we could let the browser scale it back again... I'd also consider making the image a link to the ratings site

Reply:
When you get a chance, could you turn your fork into a branch and I'll have a play (I've invited you as a collaborator)

Reply:
Ok pushed to that branch. Should be back at about 6 or so, so can take a look then :)

Reply:
FYI - this little snippet can download all the FSA open data sets in less than 30 seconds

```
curl -L https://ratings.food.gov.uk/open-data/en-gb | grep -Eo 'http://ratings.food.gov.uk/OpenDataFiles/.*\.xml' | xargs -P20 -n1 -I{} curl -LO {}
```

Reply:
Nice I'll give that a go. I was using the JSON data but imagine the XML isn't too similar. Thanks!

Reply:
I’ll have a look at JSON too - makes it easier to preprocess with jq - rather than having to faff with xpath

Reply:
I got the script working with XML. To be fair PHP handles JSON and XML in similar ways for the XML format which has been used.

Reply:
I've been playing with python for the parsing - playing with Levenshtein distances to try to get a better matching by name as well as xslt to turn the xml into CSV which plays quite neat I think. I'll push that up later.

Reply:
I found that the JSON provided by the FSA is broken, as it had some invalid escapes something like `... \& ....` which made `jq` fall over, so I've stuck to XML

Reply:
I found Levenshtein terrible for matching. It's amazing how far out some of these can be. I've just found a way to match some more. Interested to see another way though :)

Reply:
Some bad examples as follows
- Cobbies inn - DD6 9BG - FSA data  (actually appears twice as well) / Kitchinn DD6 9BG -HMRC data    Google can link these as the same place but nothing about the names would. The way im matching this doesnt match due to having 2 on the same postcode (in reality its just one the FSA data shouldnt duplicate places) 
-  The Cockney Leek Bistro  - No postcode - FSA data  - The Cockney Leek Bistro CH65 7EU  - As theres no postcode thats currently the main "link" 
- Franchises Example KFC, loads have the name in HMRC as actual name t/a KFC, FSA has KFC detail W2 1RL - FSA is "Kentucky Fried Chicken Express" HMRC is "Triple D Limited t/a KFC"

Reply:
My internet was playing up last night, so didn't get as far as I wanted - but I'm essentially using the following:

- for all items with the same postcode
- all lowercase, clear out non [a-z] characters
- sort words in alphabetical order
- filter out any word lists that don't have matching blocks of 75% the length of the biggest word
- sort by Levenshtein distance

just running a comparison to see whether that's better or worse than the existing algo.